cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java

```diff
@@ -156,53 +156,48 @@ public ProgramController run(final Program program, ProgramOptions options) {
     final Service mapReduceRuntimeService = new MapReduceRuntimeService(cConf, hConf, mapReduce, spec, context,
                                                                         program.getJarLocation(), locationFactory,
                                                                         streamAdmin, txSystemClient);
-    final ProgramController controller = new MapReduceProgramController(mapReduceRuntimeService, context);
-    controller.addListener(new AbstractListener() {
+    mapReduceRuntimeService.addListener(new Service.Listener() {
       @Override
-      public void init(ProgramController.State state, @Nullable Throwable cause) {
+      public void starting() {
         store.setStart(program.getId(), runId.getId(), TimeUnit.SECONDS.convert(System.currentTimeMillis(),
                                                                                 TimeUnit.MILLISECONDS));
-        if (state == ProgramController.State.COMPLETED) {
-          completed();
-        }
-        if (state == ProgramController.State.ERROR) {
-          error(controller.getFailureCause());
-        }
       }
 
       @Override
-      public void completed() {
-        store.setStop(program.getId(), runId.getId(),
-                      TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS),
-                      ProgramController.State.COMPLETED.getRunStatus());
+      public void running() {
```

Comment:
You can extend from `ServiceListenerAdapter` to simplify implementation for no-op method.